### PR TITLE
feat: `CofiniteIntSet`

### PR DESCRIPTION
**Pull request by JonathanStarup**
Needed for the new solver

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/unification/CofiniteIntSet.scala [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  *
+  * All sets are either a [[SortedSet]] or a complement of it.
+  *
+  * No finite set is ever equivalent to universe.
+  */
+sealed trait CofiniteIntSet {
+
+  import CofiniteIntSet.{Compl, Set}
+
+  /** Returns `true` if `this` is [[CofiniteIntSet.universe]]. */
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/CofiniteIntSet.scala:32]: Why are all your comments on 1 line? They should be formatted on multiple lines :)

File: main/src/ca/uwaterloo/flix/language/phase/unification/CofiniteIntSet.scala [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    case Compl(_) => false
+  }
+
+  /** Returns `true` if `this` is [[CofiniteIntSet.empty]]. */
+  def isEmpty: Boolean = this match {
+    case Set(s) if s.isEmpty => true
+    case Set(_) => false
+    case Compl(_) => false
+  }
+
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/CofiniteIntSet.scala:45]: I think repeating " No finite set is ever equivalent to universe." would be good for both isUniverse and isEmpty.

File: main/src/ca/uwaterloo/flix/language/phase/unification/CofiniteIntSet.scala [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  /** Returns `true` if `this` is [[CofiniteIntSet.universe]]. */
+  def isUniverse: Boolean = this match {
+    case Compl(s) if s.isEmpty => true
+    case Set(_) => false
+    case Compl(_) => false
+  }
+
+  /** Returns `true` if `this` is [[CofiniteIntSet.empty]]. */
+  def isEmpty: Boolean = this match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/CofiniteIntSet.scala:40]: Can we have isEmpy before isUniverse, as a matter of style?

File: main/src/ca/uwaterloo/flix/language/phase/unification/CofiniteIntSet.scala [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    case (Set(x), Set(y)) =>
+      // x ∪ y
+      Set(x.union(y))
+    case (Set(x), Compl(y)) =>
+      // x ∪ !y
+      // = !!(x ∪ !y)    (double complement)
+      // = !(!x ∩ y)     (complement distribution)
+      // = !(y ∩ !x)     (intersection symmetry)
+      // = !(y - x)      (difference definition)
+      Compl(y.diff(x))
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/CofiniteIntSet.scala:79]: I appreciate the comments 👍

File: main/src/ca/uwaterloo/flix/language/phase/unification/CofiniteIntSet.scala [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+      Set(y.diff(x))
+    case (Compl(x), Compl(y)) =>
+      // !x ∩ !y
+      // = !!(!x ∩ !y)   (double complement)
+      // = !(x ∪ y)      (complement distribution)
+      Compl(x.union(y))
+  }
+
+  /** Returns the complement of `s` (`!s`). */
+  def complement(s: CofiniteIntSet): CofiniteIntSet = s match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/CofiniteIntSet.scala:116]: Can we move this function before union, as a matter of style. Thanks!

File: main/src/ca/uwaterloo/flix/language/phase/unification/CofiniteIntSet.scala [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+      // !s
+      Compl(s)
+    case Compl(s) =>
+      // !!s
+      // = s             (double negation)
+      Set(s)
+  }
+
+  /**
+    * Instantiates the elements of `s` according to `m`.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/CofiniteIntSet.scala:127]: I don't fully understand what this does.

First I thought it was just map, but its not.

I think an example in the comment would be nice.

Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/unification/CofiniteIntSet.scala]: I will just not include it in this PR, it doesn't make sense without the external context

File: main/src/ca/uwaterloo/flix/language/phase/unification/CofiniteIntSet.scala (hunk 1 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/unification/CofiniteIntSet.scala]: Yeah, but we don't do that anywhere else...

**Review by magnus-madsen — APPROVED** [review]

File: main/src/ca/uwaterloo/flix/language/phase/unification/CofiniteIntSet.scala (hunk 1 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/unification/CofiniteIntSet.scala]: WIll Intellij not screw up the formatting when I open the file?

**Comment by magnus-madsen** [comment]
Do you want any tests?

File: main/src/ca/uwaterloo/flix/language/phase/unification/CofiniteIntSet.scala (hunk 1 above)
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/unification/CofiniteIntSet.scala]: No intellij also likes it

**Pull request merged**